### PR TITLE
`DigestAuthProvider` should be loaded lazily

### LINKED ISSUE #210: `DigestAuthProvider` should be loaded lazily
See https://github.com/bitfireAT/davx5/issues/937.

Proposal:

Since `digestAuthProvider` is the culprit and is used far less often than `basicAuthProvider`, we should create its instance lazily.

- [x] ideally add test that reproduces the original issue (loading `DigestAuthProvider` with congested default dispatcher can cause deadlock)
- [x] lazy-construct `digestAuthProvider`
- [x] `isApplicable`: check for `basicAuthProvider.isApplicable(auth)` first – for servers which accept Digest auth, `digestAuthProvider` won't be loaded then, not even to check whether it's applicable

### PR DESCRIPTION
**Pull request by ArnyminerZ**
- Added a test (`DigestAuthProviderCongestionTest`) which verifies that `PreemptiveBasicDigestAuthProvider` can be constructed even when the default pool is full.
  _Kept on a different class than `PreemptiveBasicDigestAuthProvider` because the test relies on `DigestAuthProvider` not being called ever before, so keeping it apart makes it more "probable"_
- Made `digestAuthProvider` initialize lazily, and added an explanatory comment for the reasoning.
- Switched `basicAuthProvider` and `digestAuthProvider` checks so that basic runs first, and `digestAuthProvider` is only used if necessary.

### CODE REVIEW

File: src/test/kotlin/at/bitfire/dav4jvm/ktor/DigestAuthProviderCongestionTest.kt [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+import org.junit.Assert.assertFalse
+import org.junit.Test
+
+/**
+ * Kept in its own class: the reproduction only works reliably as the first thing in the JVM to
+ * construct a `DigestAuthProvider`, since ktor's nonce buffer is a process-wide singleton that
+ * other tests can warm up. Run in isolation, e.g. `./gradlew test --tests "*.DigestAuthProviderCongestionTest"`.
```
Comment (rfc2822) [inline src/test/kotlin/at/bitfire/dav4jvm/ktor/DigestAuthProviderCongestionTest.kt]: **issue (non-blocking)**

It's a normal test class, so it's not guaranteed to be run before other tests.

We could add a note that we currently don't test `DigestAuthProvider` anywhere else and then it should be fine.

Also I'd like if we could reduce the test time to a minimum, maybe 0.5 sec instead of 2 sec (should be enough to instantiate a class) and explicitly stop the allCongested coroutines, for instance with a test-wide scope that is cancelled at the end.

File: src/main/kotlin/at/bitfire/dav4jvm/ktor/PreemptiveBasicDigestAuthProvider.kt [hunk 2]
```diff
@@ -40,9 +40,17 @@ class PreemptiveBasicDigestAuthProvider(
...
-    private val digestAuthProvider = DigestAuthProvider(
-        credentials = { DigestAuthCredentials(username, password) }
-    )
+    private val digestAuthProvider by lazy {
+        // DigestAuthProvider uses generateNonceBlocking, which starts a coroutine on Dispatchers.Default and then
+        // blocks until it receives a value from that coroutine. If every thread in the (CPU-core-sized)
+        // Dispatchers.Default pool is already busy, the nonce generator has nowhere to run, and construction stalls
+        // until a thread frees up. See DigestAuthProviderCongestionTest for a test that reproduces this.
+        // Initializing lazily makes sure this class can be constructed without issues.
+        // Then, in isApplicable, we only use it if necessary (basicAuthProvider is not applicable)
```
Comment (rfc2822) [inline src/main/kotlin/at/bitfire/dav4jvm/ktor/PreemptiveBasicDigestAuthProvider.kt:46]: **Todo:** I'd add the actual Ktor bug URL (we have to report it before) and then keep the comment as short as possible.

Even without the bug, lazy-initializing the digest provider is a good idea because it's rarely used.

File: src/test/kotlin/at/bitfire/dav4jvm/ktor/DigestAuthProviderCongestionTest.kt [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+ */
+class DigestAuthProviderCongestionTest {
+
+    @Test
+    @OptIn(DelicateCoroutinesApi::class)
+    fun `construction does not block when Dispatchers-Default is congested`() {
```
Comment (rfc2822) [inline src/test/kotlin/at/bitfire/dav4jvm/ktor/DigestAuthProviderCongestionTest.kt:33]: **Todo:**

- Once we have the Ktor bug URL, we should also add it here.
- We should note that as soon as the test starts to fail, we can remove it and all references to the Ktor bug (but still keep the digest authenticator lazy-loaded for performance reasons).

**Review by rfc2822 — COMMENTED** [review]
Nice! I think we should report the issue to Ktor and include the URL and then this is ready.

**Comment by ArnyminerZ** [comment]
Reported to Ktor: https://youtrack.jetbrains.com/issue/KTOR-9722/DigestAuthProvider-cannot-be-initialized-with-a-congested-Dispatchers.Default-pool

**Review by rfc2822 — APPROVED** [review]
Looks good!

**Comment by rfc2822** [comment]
We should not forget to update dav4jvm in DAVx5 then too.

And probably we have to finish / merge https://github.com/bitfireAT/davx5-ose/pull/2640 first
